tools/cli/adaparse.cpp

```diff
@@ -0,0 +1,287 @@
+#endif
+
+uint64_t nano() {
+  return std::chrono::duration_cast<::std::chrono::nanoseconds>(
+             std::chrono::steady_clock::now().time_since_epoch())
+      .count();
+}
+
+template <typename Callable>
+bool print_part(Callable&& fmt_or_adaparse_print, std::string_view get_part,
+                const ada::url_aggregator& url) {
+  if (get_part.size() == 4) {
+    if (get_part[0] == 'h') {
+      if (get_part == "host") {
+        fmt_or_adaparse_print("{}\n", url.get_host());
+        return true;
+      } else if (get_part == "hash") {
+        fmt_or_adaparse_print("{}\n", url.get_hash());
+        return true;
+      }
+    } else if (get_part[0] == 'p') {
+      if (get_part == "port") {
+        fmt_or_adaparse_print("{}\n", url.get_port());
+        return true;
+      }
+    }
+  } else if (get_part.size() == 6) {
+    if (get_part == "origin") {
+      fmt_or_adaparse_print("{}\n", url.get_origin());
+      return true;
+    }
+    if (get_part == "search") {
+      fmt_or_adaparse_print("{}\n", url.get_search());
+      return true;
+    }
+  } else if (get_part.size() == 8) {
+    if (get_part[0] == 'p') {
+      if (get_part == "protocol") {
+        fmt_or_adaparse_print("{}\n", url.get_protocol());
+        return true;
+      }
+      if (get_part == "password") {
+        fmt_or_adaparse_print("{}\n", url.get_password());
+        return true;
+      }
+      if (get_part == "pathname") {
+        fmt_or_adaparse_print("{}\n", url.get_pathname());
+        return true;
+      }
+    }
+  } else if (get_part == "hostname") {
+    fmt_or_adaparse_print("{}\n", url.get_hostname());
+    return true;
+
+  } else if (get_part == "username") {
+    fmt_or_adaparse_print("{}\n", url.get_username());
+    return true;
+  }
+
+  fmt::print(stderr, "\"{}\" not found\n", get_part);
+  return false;
+}
+
+// This function parses a FILE * descriptor, line by line (URL by URL).It
+// applies the given Callable (in this case the lambda adaparse_print in the
+// main() function) to print the output either to a file on disk or to the
+// console, depending on arguments given.
+// It also optionally will output a benchmark to the console.
+template <typename Callable>
+int piped_file(Callable&& adaparse_print, const cxxopts::ParseResult result,
+               FILE* input_file) {
+  constexpr size_t cache_length = 32768;
+  std::unique_ptr<char[]> cachebuffer(new char[cache_length]{});
+
+  uint64_t before = nano();
+
+  size_t total_bytes_read = 0;
+  size_t bytes_read_this_loop_iteration;
+  size_t offset = 0;
+  size_t lines = 0;
+  size_t sum_of_lines_bytes = 0;
+  size_t blocks = 0;
+
+  std::string get_part;
+  if (result.count("get")) {
+    get_part = result["get"].as<std::string>();
+  }
+
+  // Get the file descriptor from the FILE * input_file
+  int input_fd = fileno(input_file);
+
+  while ((bytes_read_this_loop_iteration = read(
+              input_fd, cachebuffer.get() + offset, cache_length - offset))) {
```

Comment:
size_t needs to be casted properly in here. this is the cause of msvc fail @lemire

Reply:
```
  D:\a\ada\ada\tools\cli\adaparse.cpp(104,1): warning C4267: 'argument': conversion from 'size_t' to 'unsigned int', possible loss of data [D:\a\ada\ada\build\tools\cli\adaparse.vcxproj]
```